sql: support AS OF SYSTEM TIME readonly queries

madelynnblue opened the pull request:
This is implemented by adding special logic to the executor. It is
unfortunate that we have to teach it so much about how SELECT statements
work, but the executor is the thing that can set the proto timestamps
during each retry attempt (they cannot only be set once since the
auto-retry stuff will bump the timestamps otherwise).

In the case of an AS OF query, we only want to fetch the table descriptor
at that time, and not a lease, even if the time is current.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7139)

<!-- Reviewable:end -->

tschottdorf commented:
Looks good generally.

<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> #### sql: support AS OF SYSTEM TIME readonly queries
> 
> This is implemented by adding special logic to the executor. It is
> 
> unfortunate that we have to teach it so much about how SELECT statements
> 
> work, but the executor is the thing that can set the proto timestamps
> 
> during each retry attempt (they cannot only be set once since the
> 
> auto-retry stuff will bump the timestamps otherwise).
> 
> 
> 
> In the case of an AS OF query, we only want to fetch the table descriptor
> 
> at that time, and not a lease, even if the time is current.
> 
> 
> 
</details>


---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sql/as_of_test.go, line 50 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqTOHZjOGCwIC8ocQI:-KJqTOHZjOGCwIC8ocQJ:-1056372613) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L50)):*
> ```Go
> 	var n1, n2 time.Time
> 
> 	// Expect an error if table doesn't exist at time.
> ```

I don't understand this comment.

---

*[sql/as_of_test.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqTTZXYAt6NAtwFvnM:-KJqTTZXYAt6NAtwFvnN:-778865526) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L57)):*
> ```Go
> 	if _, err := db.Query(fmt.Sprintf("SELECT a FROM d.t AS OF SYSTEM TIME '%s'", d.String())); err == nil {
> 		t.Fatal("expected error")
> 	} else if err.Error() != `pq: database "d" does not exist` {
> ```

Use `testutils.IsError()`.

It would also be interesting to run this select after you create the table (but you should still get an error because the db/table did not exist at n1). Ideally testing both cases (db and table).

Can you sprinkle commentary through this test and avoid overwriting variables (instead naming them better? Something like tsPreCreate, tsPostCreate, tsPostInsert, tsPostUpdate, ...)

---

*[sql/as_of_test.go, line 126 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqUbeKtkzv79DYUmrv:-KJqUbeLX6lu3ohz0aGZ:1676697132) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L126)):*
> ```Go
> 	if err := db.QueryRow("SELECT a FROM d.t").Scan(&i); err == nil {
> 		t.Fatal("expected error")
> 	} else if err.Error() != `pq: qualified name "a" not found` {
> ```

`testutils.IsError`

---

*[sql/as_of_test.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqUpWgxuPvBtQUZlkU:-KJqUpWgxuPvBtQUZlkV:-883853797) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L162)):*
> ```Go
> 	}
> 
> 	var m int64
> ```

Why not name this and everything else here more descriptively?

---

*[sql/as_of_test.go, line 188 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqV-69wLFJYpslv-Bo:-KJqV-69wLFJYpslv-Bp:-1813541893) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L188)):*
> ```Go
> 				if count > 0 && bytes.Contains(req.Key, []byte(key)) {
> 					magicVals.restartCounts[key]--
> 					err := roachpb.NewTransactionPushError(*hdr.Txn)
> ```

TransactionPushError isn't what you want. You want a `TransactionRetryError` carrying a copy of the incoming transaction whose timestamp has been pushed (I suggest pushing the logical component only to avoid funny business), and asserting here that the incoming transaction always has as the correct deadline (even if you let it commit).

---

*[sql/executor.go, line 407 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqWbu969XXW9dXNX9e:-KJqWbu969XXW9dXNX9f:1457419251) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/executor.go#L407)):*
> ```Go
> 		// can be consumed.
> 		stmtsToExec := stmts
> 		var protoTS *roachpb.Timestamp
> ```

Add comment.

---

*[sql/executor.go, line 461 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqWi_0eRxVDzb8fr14:-KJqWi_0eRxVDzb8fr15:1655501429) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/executor.go#L461)):*
> ```Go
> 			if protoTS != nil {
> 				txnState.txn.Proto.OrigTimestamp = *protoTS
> 				txnState.txn.Proto.MaxTimestamp = *protoTS
> ```

You need to set the deadline. Without this, the suggested updates to the test above should expose this as a bug.

---

*[sql/planner.go, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqXC2HIIpFxTbbe5z2:-KJqXC2HIIpFxTbbe5z3:-137594938) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/planner.go#L58)):*
> ```Go
> 	// have already been set correctly, and this flag is used in layers below
> 	// the executor to modify the behavior of the SELECT. For example the table
> 	// descriptor is not leased, only fetched at the correct time.
> ```

So there's no caching of the table descriptor? That sounds like something we're going to have to track (and eventually fix). Could you file it before this merges?

---

*[sql/scan.go, line 253 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqXMIuxiKimAvw3AB7:-KJqXMIuxiKimAvw3AB8:188913962) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/scan.go#L253)):*
> ```Go
> 	var err error
> 
> 	if p.asOf {
> ```

A comment wouldn't hurt.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-:-KJqXbKM6SY0IElbvWQG:-1151933653)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 7 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[sql/as_of_test.go, line 50 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqTOHZjOGCwIC8ocQI:-KJqb6kmXTuflazEdHsd:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L50)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I don't understand this comment.
</details>
Done.

---

*[sql/as_of_test.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqTTZXYAt6NAtwFvnM:-KJqdiOqGbj7yxKB3qpy:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L57)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Use `testutils.IsError()`.
> 
> It would also be interesting to run this select after you create the table (but you should still get an error because the db/table did not exist at n1). Ideally testing both cases (db and table).
> 
> Can you sprinkle commentary through this test and avoid overwriting variables (instead naming them better? Something like tsPreCreate, tsPostCreate, tsPostInsert, tsPostUpdate, ...)
</details>
Done.

---

*[sql/as_of_test.go, line 126 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqUbeKtkzv79DYUmrv:-KJqbfVV0kOTi6-IsWfQ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L126)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `testutils.IsError`
</details>
Done.

---

*[sql/as_of_test.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqUpWgxuPvBtQUZlkU:-KJqfAPHuzImoHz9TiJI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L162)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Why not name this and everything else here more descriptively?
</details>
Done.

---

*[sql/as_of_test.go, line 188 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqV-69wLFJYpslv-Bo:-KJr0l9LUlPK_K1iVqX_:-779093239) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L188)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> TransactionPushError isn't what you want. You want a `TransactionRetryError` carrying a copy of the incoming transaction whose timestamp has been pushed (I suggest pushing the logical component only to avoid funny business), and asserting here that the incoming transaction always has as the correct deadline (even if you let it commit).
</details>
I have changed the error to this type but the deadline bugs you say should happen are not. I'm not sure what you mean exactly by "carrying a copy". `NewErrorWithTxn` is called a few lines down which is what I think you mean. If that is correct, then I'm not sure why the deadline error isn't triggering.

---

*[sql/executor.go, line 407 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqWbu969XXW9dXNX9e:-KJqeNYty7CX9Yw0MH2I:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/executor.go#L407)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Add comment.
</details>
Done.

---

*[sql/planner.go, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqXC2HIIpFxTbbe5z2:-KJqf5lQmvKGWe4QzrwP:-1844941681) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/planner.go#L58)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> So there's no caching of the table descriptor? That sounds like something we're going to have to track (and eventually fix). Could you file it before this merges?
</details>
Filed #7143.

---

*[sql/scan.go, line 253 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqXMIuxiKimAvw3AB7:-KJqemUy28IKSp2wSZua:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/scan.go#L253)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> A comment wouldn't hurt.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 10 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[sql/executor.go, line 1173 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJrItyf5HGM7MnfsIxe:-KJrItyf5HGM7MnfsIxf:257082369) ([raw file](https://github.com/cockroachdb/cockroach/blob/498b719abf6391d2c59e8f3d08dd654990a19fc3/sql/executor.go#L1173)):*
> ```Go
> 
> func isAsOf(planMaker *planner, stmt parser.Statement, max roachpb.Timestamp) (*roachpb.Timestamp, error) {
> 	s, ok := stmt.(*parser.Select)
> ```

I'd suggest adding a comment here "Analysis of the select statement is done here to bypass the logic in newPlan(), since that requires the transaction to be started already."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[sql/as_of_test.go, line 188 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqV-69wLFJYpslv-Bo:-KJs4UMBpUNndL4rf4YO:-890563736) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L188)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> I have changed the error to this type but the deadline bugs you say should happen are not. I'm not sure what you mean exactly by "carrying a copy". `NewErrorWithTxn` is called a few lines down which is what I think you mean. If that is correct, then I'm not sure why the deadline error isn't triggering.
</details>
You're using a `TransactionRetryError`, but you're not doing the rest (populating it with a txn that carries a higher timestamp). You must make a copy of `args.Hdr.Txn` and modify it as suggested before returning it with the error - you're currently returning the same one, so no increased timestamp.

---

*[sql/as_of_test.go, line 104 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJs3RTEXB9b6IOpbtx3:-KJs3RTEXB9b6IOpbtx4:2057463158) ([raw file](https://github.com/cockroachdb/cockroach/blob/498b719abf6391d2c59e8f3d08dd654990a19fc3/sql/as_of_test.go#L104)):*
> ```Go
> 		t.Fatalf("expected %v, got %v", val1, i)
> 	}
> 	if err := db.QueryRow("SELECT a FROM d.t").Scan(&i); err != nil {
> ```

This is already done above. Is there a particular reason to have this twice?

---

*[sql/as_of_test.go, line 117 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJs3m7X3HyhFVl39Ged:-KJs3m7X3HyhFVl39Gee:1390997042) ([raw file](https://github.com/cockroachdb/cockroach/blob/498b719abf6391d2c59e8f3d08dd654990a19fc3/sql/as_of_test.go#L117)):*
> ```Go
> 	}
> 
> 	// Future queries shouldn't work.
> ```

What about reading at a past timestamp which has since been GC'ed? Probably needs a little more setup (I think writing a timestamp into the last-gc'ed field when the first range is bootstrapped is enough, given that we copy that information on a split (right?)); then you should be able to use a time like '1970-01-01' to trigger the error. Just to make sure that it fails in the way that we think. Maybe that error needs to be converted to some error prescribed by the SQL standard, too (if an adequate counterpart exists).

---

*[sql/as_of_test.go, line 140 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJs3gm3mnM1uQXuVYhT:-KJs3gm3mnM1uQXuVYhU:131192123) ([raw file](https://github.com/cockroachdb/cockroach/blob/498b719abf6391d2c59e8f3d08dd654990a19fc3/sql/as_of_test.go#L140)):*
> ```Go
> 		t.Fatalf("expected %v, got %v", val1, i)
> 	}
> 	if err := db.QueryRow("SELECT a FROM d.t").Scan(&i); err == nil {
> ```

What does this test?

---

*[sql/as_of_test.go, line 178 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJs5ijeVeqsr4fL-uBn:-KJs5ijfZVItvVbyqbBM:329425513) ([raw file](https://github.com/cockroachdb/cockroach/blob/498b719abf6391d2c59e8f3d08dd654990a19fc3/sql/as_of_test.go#L178)):*
> ```Go
> 	}
> 
> 	var logical int64
> ```

I think a better name for this would be `walltime`. `logical` makes me think of the logical portion of the HLC timestamps, but that is precisely what you're throwing away here (by casting the decimal to an int).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 8 of 10 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[sql/as_of_test.go, line 104 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJs3RTEXB9b6IOpbtx3:-KJtP3KYc_FkxU40dDyc:1818235899) ([raw file](https://github.com/cockroachdb/cockroach/blob/498b719abf6391d2c59e8f3d08dd654990a19fc3/sql/as_of_test.go#L104)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This is already done above. Is there a particular reason to have this twice?
</details>
Nope. Removed.

---

*[sql/as_of_test.go, line 117 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJs3m7X3HyhFVl39Ged:-KK00AJR5Hm_QeYbP3cK:1578807197) ([raw file](https://github.com/cockroachdb/cockroach/blob/498b719abf6391d2c59e8f3d08dd654990a19fc3/sql/as_of_test.go#L117)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What about reading at a past timestamp which has since been GC'ed? Probably needs a little more setup (I think writing a timestamp into the last-gc'ed field when the first range is bootstrapped is enough, given that we copy that information on a split (right?)); then you should be able to use a time like '1970-01-01' to trigger the error. Just to make sure that it fails in the way that we think. Maybe that error needs to be converted to some error prescribed by the SQL standard, too (if an adequate counterpart exists).
</details>
Instead of triggering a GC, I'm using a pre-1970 timestamp, which appears to work.

---

*[sql/as_of_test.go, line 140 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJs3gm3mnM1uQXuVYhT:-KJtPaO0N2kJFZB7chd9:1684872152) ([raw file](https://github.com/cockroachdb/cockroach/blob/498b719abf6391d2c59e8f3d08dd654990a19fc3/sql/as_of_test.go#L140)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What does this test?
</details>
My thought was to verify that we get that expected error in the present after the other expected error from the past. Not strictly needed I suppose. Removed.

---

*[sql/as_of_test.go, line 178 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJs5ijeVeqsr4fL-uBn:-KJtQ1GUV6dIeG9hSx-O:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/498b719abf6391d2c59e8f3d08dd654990a19fc3/sql/as_of_test.go#L178)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I think a better name for this would be `walltime`. `logical` makes me think of the logical portion of the HLC timestamps, but that is precisely what you're throwing away here (by casting the decimal to an int).
</details>
Done.

---

*[sql/executor.go, line 461 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqWi_0eRxVDzb8fr14:-KK-wijxu0OPdx-tUYyZ:-2030534715) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/executor.go#L461)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> You need to set the deadline. Without this, the suggested updates to the test above should expose this as a bug.
</details>
I did a bunch of tests and am not convinced about this. The deadline checking examines the `Timestamp` field of the proto. Since (as of this most recent revision) I am setting `.Timestamp` at the start of each retry, I do not see how it could ever trigger the deadline failure. Indeed, with the modifications to the test suggested above, commenting out the deadline setting in executor.go doesn't change the output of the test. I have left the deadline setting in, because I don't think it does any harm and could potentially find problems in the future. But I'm not sure it's doing anything explicitly useful right now.

---

*[sql/executor.go, line 1173 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJrItyf5HGM7MnfsIxe:-KJtQXzAJki1mlYrLGRb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/498b719abf6391d2c59e8f3d08dd654990a19fc3/sql/executor.go#L1173)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I'd suggest adding a comment here "Analysis of the select statement is done here to bypass the logic in newPlan(), since that requires the transaction to be started already."
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 8 of 10 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[sql/as_of_test.go, line 188 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqV-69wLFJYpslv-Bo:-KK0814IVQLEh6mNJM9Z:-1001325554) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/as_of_test.go#L188)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> You're using a `TransactionRetryError`, but you're not doing the rest (populating it with a txn that carries a higher timestamp). You must make a copy of `args.Hdr.Txn` and modify it as suggested before returning it with the error - you're currently returning the same one, so no increased timestamp.
</details>
Done, I think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 8 of 10 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 421 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA9R4qBFspOTt_dCdc:-KKA9R4ry6DIOE_VhNDH:-1263858393) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L421)):*
> ```Go
> 				// Check for AS OF SYSTEM TIME. If it is present but not detected here,
> 				// it will raise an error later on.
> 				protoTS, err = isAsOf(planMaker, stmtsToExec[0], execOpt.MinInitialTimestamp)
> ```

how about `txnTimestamp`?

---

*[sql/executor.go, line 421 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA9a8jUDQDE4_2dhOv:-KKA9a8jUDQDE4_2dhOw:-1163371514) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L421)):*
> ```Go
> 				// Check for AS OF SYSTEM TIME. If it is present but not detected here,
> 				// it will raise an error later on.
> 				protoTS, err = isAsOf(planMaker, stmtsToExec[0], execOpt.MinInitialTimestamp)
> ```

is there a point in supporting txn retries for these reads, given that we can't update the txn timestamp?
How about verifying that we're below the uncertainty period, and then asserting that there's not retries?

---

*[sql/executor.go, line 465 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA9F_-OG0Pnwdkvsf8:-KKA9F_0SxRG-MfYjadU:2017320058) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L465)):*
> ```Go
> 				txnState.txn.Proto.OrigTimestamp = *protoTS
> 				txnState.txn.Proto.MaxTimestamp = *protoTS
> 				txnState.txn.UpdateDeadlineMaybe(*protoTS)
> ```

assert on the result of `UpdateDeadlineMaybe` (or introduce a result if there isn't one). Cause this code doesn't like the `Maybe` aspect :)

---

*[sql/executor.go, line 1176 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA7g15p7nGStl6wokK:-KKA7g16mUOwwQJ8LLQp:-973098702) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L1176)):*
> ```Go
> // isAsOf analyzes a select statement to bypass the login in newPlan(),
> // since that requires the transaction to be started already.
> func isAsOf(planMaker *planner, stmt parser.Statement, max roachpb.Timestamp) (*roachpb.Timestamp, error) {
> ```

please comment the return val

---

*[sql/executor.go, line 1217 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA8ojdHIKmjt7Nxv95:-KKA8ojdHIKmjt7Nxv96:1769442359) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L1217)):*
> ```Go
> 	}
> 	if max.Less(ts) {
> 		return nil, fmt.Errorf("cannot specify timestamp in the future")
> ```

consider add an `ErrorWithPGCode`, otherwise this will result in an internal error being returned to the client

---

*[sql/planner.go, line 56 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA88Asr70SnaXq-joH:-KKA88At0TEUCnm434Fa:-1211605990) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/planner.go#L56)):*
> ```Go
> 
> 	// If set, this is an AS OF SYSTEM TIME query. The transaction timestamps
> 	// have already been set correctly, and this flag is used in layers below
> ```

this comment is a bit weird. Not clear what transaction timestamps it's talking about.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 8 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 421 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA9R4qBFspOTt_dCdc:-KKAEsgtVbnL7M_NEyXd:-111832821) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L421)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how about `txnTimestamp`?
</details>
Because that uses the `PhysicalClock`, and we want the HLC clock.

---

*[sql/executor.go, line 421 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA9a8jUDQDE4_2dhOv:-KKAGduJbI4-JGnquWrh:-787130849) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L421)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> is there a point in supporting txn retries for these reads, given that we can't update the txn timestamp?
> How about verifying that we're below the uncertainty period, and then asserting that there's not retries?
</details>
I think retries can occur if there's an old intent, say, in a long-running and still open transaction.

---

*[sql/executor.go, line 465 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA9F_-OG0Pnwdkvsf8:-KKAGw1_DyUlLil5ftYZ:-2124874) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L465)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> assert on the result of `UpdateDeadlineMaybe` (or introduce a result if there isn't one). Cause this code doesn't like the `Maybe` aspect :)
</details>
I did that and it causes a bug. The function returns true if it set the deadline, which it does only if it's strictly older than the previous deadline. Thus, setting the same deadline more than 1 times will return false for all times after the first. It's not a useful thing to check: it's not an error if it's not set.

---

*[sql/executor.go, line 1176 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA7g15p7nGStl6wokK:-KKAHa6hHshMqs5-oZNC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L1176)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> please comment the return val
</details>
Done.

---

*[sql/executor.go, line 1217 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA8ojdHIKmjt7Nxv95:-KKAKB4EnFbPFZ_W9d6h:1286034962) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L1217)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> consider add an `ErrorWithPGCode`, otherwise this will result in an internal error being returned to the client
</details>
I considered it and have decided to not. The best error I could come up with is 22000, a generic "data exception", but I have no idea if that's an appropriate use of that error or not.

---

*[sql/planner.go, line 56 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA88Asr70SnaXq-joH:-KKAKld3AS0mEhUY2_rs:-104774680) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/planner.go#L56)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> this comment is a bit weird. Not clear what transaction timestamps it's talking about.
</details>
Agree. I removed the note about transaction timestamps since it's not needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Looks good generally.
</details>


---

Review status: 3 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 421 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA9a8jUDQDE4_2dhOv:-KKANBa0nccT94to_tfs:-673227915) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L421)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> I think retries can occur if there's an old intent, say, in a long-running and still open transaction.
</details>
ah, yes, old intents. Although we saw that the read would actually get blocked and wait for those intents with the current code. In any case, I would think it's OK to just fail the historical read query in that case saying "history is still being written" and not bother with retries? Or maybe completely ignore intents that cannot be resolved, if that's a thing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-:-KKAO9o1e51c-DGglBBp:676834233)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 3 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[sql/executor.go, line 421 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA9a8jUDQDE4_2dhOv:-KKAV9omDExo-YLaMQ3v:-208889417) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L421)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> ah, yes, old intents. Although we saw that the read would actually get blocked and wait for those intents with the current code. In any case, I would think it's OK to just fail the historical read query in that case saying "history is still being written" and not bother with retries? Or maybe completely ignore intents that cannot be resolved, if that's a thing?
</details>
I'm wrong. I forgot that they blocked lower down. In that case I am not sure what could theoretically cause a retry. In the case of old intents, it still sits there and blocks. That behavior is ok for now, as a user can ctrl+c out if the want. I agree we should add some kind of error or flag the store to not spin or something for this case.

Toby may know what can cause a retry to occur.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 3 of 10 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[sql/as_of_test.go, line 56 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB6wyQD6Uu1w1WrZH3:-KKB6wyQD6Uu1w1WrZH4:-1936806374) ([raw file](https://github.com/cockroachdb/cockroach/blob/3de3bbab5b5c0b59f9a1acf16567e6c59a5a4052/sql/as_of_test.go#L56)):*
> ```Go
> 	}
> 	tsEmpty := tm.Format(time.RFC3339Nano)
> 	if _, err := db.Query(fmt.Sprintf("SELECT a FROM d.t AS OF SYSTEM TIME '%s'", tsEmpty)); !testutils.IsError(err, `pq: database "d" does not exist`) {
> ```

Can we support placeholders here? As soon as #7172 is merged `safesql` should start complaining about this. 

---

*[sql/as_of_test.go, line 149 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB8po7iNBKrZc1mYhp:-KKB8po7iNBKrZc1mYhq:2044946068) ([raw file](https://github.com/cockroachdb/cockroach/blob/3de3bbab5b5c0b59f9a1acf16567e6c59a5a4052/sql/as_of_test.go#L149)):*
> ```Go
> 		t.Fatal("unexpected error:", err)
> 	}
> }
> ```

Add a test for `SELECT now() AS OF SYSTEM TIME '...'`. (It returns the "as of" time and not the real time, right?)

It would be nice to support expressions like `AS OF SYSTEM TIME now() - interval '1h'` but it makes my head hurt to think about `now()` evaluating differently in different parts of the same statement. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 3 of 10 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[sql/as_of_test.go, line 56 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB6wyQD6Uu1w1WrZH3:-KKCmK3L6P2Pg7Is21wV:1307699872) ([raw file](https://github.com/cockroachdb/cockroach/blob/3de3bbab5b5c0b59f9a1acf16567e6c59a5a4052/sql/as_of_test.go#L56)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Can we support placeholders here? As soon as #7172 is merged `safesql` should start complaining about this. 
</details>
No, we can't support placeholders. The query needs to be type-checked based on the table descriptor at the "as of" time. Type check is performed during the prepare phase to infer the placeholder types, and thus needs access to a table descriptor. But since we don't know the time at which we need the table descriptor, the current table descriptor is leased and used. Thus, if you try to read a column in the past that has been deleted in the future (which is valid, and we have a test to do so), then that test will fail if it's using a placeholder because it looks for the column in the current table, which has been deleted.

---

*[sql/as_of_test.go, line 149 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB8po7iNBKrZc1mYhp:-KKCMQMeGqG0cC2b0cSB:-423785888) ([raw file](https://github.com/cockroachdb/cockroach/blob/3de3bbab5b5c0b59f9a1acf16567e6c59a5a4052/sql/as_of_test.go#L149)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Add a test for `SELECT now() AS OF SYSTEM TIME '...'`. (It returns the "as of" time and not the real time, right?)
> 
> It would be nice to support expressions like `AS OF SYSTEM TIME now() - interval '1h'` but it makes my head hurt to think about `now()` evaluating differently in different parts of the same statement. 
</details>
That would return the real time, not the as of time, which is currently by design. I think that's correct because the syntax implies that we are querying a table at a specific time, not setting the whole statement to a time. If the syntax were something like `BEGIN READONLY AS OF <time>`, then I think the time functions should return the as of time. Agree?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 3 of 10 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[sql/as_of_test.go, line 56 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB6wyQD6Uu1w1WrZH3:-KKCy02lRp9wJ4Zd5pFv:1530127566) ([raw file](https://github.com/cockroachdb/cockroach/blob/016ba68248e40350321a99c8f495a160a823812a/sql/as_of_test.go#L56)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> No, we can't support placeholders. The query needs to be type-checked based on the table descriptor at the "as of" time. Type check is performed during the prepare phase to infer the placeholder types, and thus needs access to a table descriptor. But since we don't know the time at which we need the table descriptor, the current table descriptor is leased and used. Thus, if you try to read a column in the past that has been deleted in the future (which is valid, and we have a test to do so), then that test will fail if it's using a placeholder because it looks for the column in the current table, which has been deleted.
</details>
Well this is not entirely true; you can invoke TypeCheck then Eval on the SYSTEM TIME expression separately; as long as there are no QualifiedNames in there there will be no error due to missing table schema. Ping me if you want details.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r3, 4 of 6 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[sql/as_of_test.go, line 149 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB8po7iNBKrZc1mYhp:-KKEZud2DAV1AoqTNN6U:767743823) ([raw file](https://github.com/cockroachdb/cockroach/blob/016ba68248e40350321a99c8f495a160a823812a/sql/as_of_test.go#L149)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> That would return the real time, not the as of time, which is currently by design. I think that's correct because the syntax implies that we are querying a table at a specific time, not setting the whole statement to a time. If the syntax were something like `BEGIN READONLY AS OF <time>`, then I think the time functions should return the as of time. Agree?
</details>
In any case, worth testing are those time functions which actually use the txn timestamp (cluster_logical_timestamp?), unless you already do so somewhere.

---

*[sql/executor.go, line 461 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqWi_0eRxVDzb8fr14:-KKE_n05IyL0XUzeLqQT:1334247943) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/executor.go#L461)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> I did a bunch of tests and am not convinced about this. The deadline checking examines the `Timestamp` field of the proto. Since (as of this most recent revision) I am setting `.Timestamp` at the start of each retry, I do not see how it could ever trigger the deadline failure. Indeed, with the modifications to the test suggested above, commenting out the deadline setting in executor.go doesn't change the output of the test. I have left the deadline setting in, because I don't think it does any harm and could potentially find problems in the future. But I'm not sure it's doing anything explicitly useful right now.
</details>
That's a good point. Could you add a comment to line 465 below?

---

*[sql/executor.go, line 421 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKA9a8jUDQDE4_2dhOv:-KKE_DR2kKVvZfJML53J:1285182918) ([raw file](https://github.com/cockroachdb/cockroach/blob/50527d638766e1e1d21d2eece9bc69ce75362844/sql/executor.go#L421)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> I'm wrong. I forgot that they blocked lower down. In that case I am not sure what could theoretically cause a retry. In the case of old intents, it still sits there and blocks. That behavior is ok for now, as a user can ctrl+c out if the want. I agree we should add some kind of error or flag the store to not spin or something for this case.
> 
> Toby may know what can cause a retry to occur.
</details>
Yes, I don't see an immediate reason for a retry to actually occur. Regardless, there should be code that handles them somewhat gracefully - we shouldn't try to second-guess the KV layer (whose behavior may change over time).

---

*[sql/executor.go, line 1174 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKE_zRPIM1zTnV5gLge:-KKE_zRPIM1zTnV5gLgf:-1705672939) ([raw file](https://github.com/cockroachdb/cockroach/blob/016ba68248e40350321a99c8f495a160a823812a/sql/executor.go#L1174)):*
> ```Go
> }
> 
> // isAsOf analyzes a select statement to bypass the login in newPlan(),
> ```

What do you mean by `login`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-:-KKEaJinXxZyy3AUBH34:1417088721)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[sql/as_of_test.go, line 56 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB6wyQD6Uu1w1WrZH3:-KKCou0Bp6D2fxmV1_3L:-1014805432) ([raw file](https://github.com/cockroachdb/cockroach/blob/016ba68248e40350321a99c8f495a160a823812a/sql/as_of_test.go#L56)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Well this is not entirely true; you can invoke TypeCheck then Eval on the SYSTEM TIME expression separately; as long as there are no QualifiedNames in there there will be no error due to missing table schema. Ping me if you want details.
</details>
What about this:

select a+$1 from t as of system time $2

It's impossible to know the type of a (and thus $1) until we know the value of $2 because a could have been dropped and added as a different type. If the system time placeholder is the only placeholder, then yes, we could delay type checking until execution time. But for now I'm going to leave it off.

---

*[sql/as_of_test.go, line 149 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB8po7iNBKrZc1mYhp:-KKEtR1etnfXRd1KSvf2:1098230143) ([raw file](https://github.com/cockroachdb/cockroach/blob/016ba68248e40350321a99c8f495a160a823812a/sql/as_of_test.go#L149)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> In any case, worth testing are those time functions which actually use the txn timestamp (cluster_logical_timestamp?), unless you already do so somewhere.
</details>
Test them for what, though? AS OF SYSTEM TIME only takes a string: you can't do `AS OF SYSTEM TIME cluster_logical_timestamp()`. Not sure what you mean here.

---

*[sql/executor.go, line 461 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KJqWi_0eRxVDzb8fr14:-KKEuQFMWenXHEMj31eF:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9477baf45c88c44e93406f5e34c452ab1228fc33/sql/executor.go#L461)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> That's a good point. Could you add a comment to line 465 below?
</details>
Done.

---

*[sql/executor.go, line 1174 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKE_zRPIM1zTnV5gLge:-KKEuZjQFBa0113btXdR:878191327) ([raw file](https://github.com/cockroachdb/cockroach/blob/016ba68248e40350321a99c8f495a160a823812a/sql/executor.go#L1174)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What do you mean by `login`?
</details>
typo. should be logic. fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 9 of 10 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[sql/as_of_test.go, line 56 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB6wyQD6Uu1w1WrZH3:-KKF3MVPbgAk5S5rD9-q:1925430665) ([raw file](https://github.com/cockroachdb/cockroach/blob/016ba68248e40350321a99c8f495a160a823812a/sql/as_of_test.go#L56)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> What about this:
> 
> select a+$1 from t as of system time $2
> 
> It's impossible to know the type of a (and thus $1) until we know the value of $2 because a could have been dropped and added as a different type. If the system time placeholder is the only placeholder, then yes, we could delay type checking until execution time. But for now I'm going to leave it off.
</details>
OK, I'm fine pushing off placeholder support here. 

---

*[sql/as_of_test.go, line 149 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB8po7iNBKrZc1mYhp:-KKF3Tk7eFAd8h1JVuW3:1069919839) ([raw file](https://github.com/cockroachdb/cockroach/blob/016ba68248e40350321a99c8f495a160a823812a/sql/as_of_test.go#L149)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Test them for what, though? AS OF SYSTEM TIME only takes a string: you can't do `AS OF SYSTEM TIME cluster_logical_timestamp()`. Not sure what you mean here.
</details>
I misremembered which time functions use which time source. I was suggesting that we have tests of `SELECT time_func() AS OF SYSTEM TIME 'T'` and verify that they do or do not use the specified time. (just to document and specify the behavior we expect here). 

What about non-literal expressions in the `AS OF SYSTEM TIME` clause? It would be nice to support them, but only if it's easy. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-:-KKFDXF4QDJ9Yh6RR8XE:676834233)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 8 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[sql/as_of_test.go, line 149 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139#-KKB8po7iNBKrZc1mYhp:-KKFEpigPLDjIuWvx87p:456849599) ([raw file](https://github.com/cockroachdb/cockroach/blob/016ba68248e40350321a99c8f495a160a823812a/sql/as_of_test.go#L149)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I misremembered which time functions use which time source. I was suggesting that we have tests of `SELECT time_func() AS OF SYSTEM TIME 'T'` and verify that they do or do not use the specified time. (just to document and specify the behavior we expect here). 
> 
> What about non-literal expressions in the `AS OF SYSTEM TIME` clause? It would be nice to support them, but only if it's easy. 
</details>
That is difficult because supporting general expressions (as long as they are of type string) require a full evaluation context to correctly run. For example, the `now()` function asserts (with a panic if it fails) that its return value is not the zero timestamp. But since this expression is being evaluated way high up in the executor, there's not a well-populated eval context yet, and so now() and the other time functions will panic because their timestamps are still the zero timestamp. Thus, we support only string literals for now because we know that they don't assume anything about the eval context.

Added tests for the time functions, though. Will merge once the circle ci tests pass.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: support AS OF SYSTEM TIME readonly queries

This is implemented by adding special logic to the executor. It is
unfortunate that we have to teach it so much about how SELECT statements
work, but the executor is the thing that can set the proto timestamps
during each retry attempt (they cannot only be set once since the
auto-retry stuff will bump the timestamps otherwise).

In the case of an AS OF query, we only want to fetch the table descriptor
at that time, and not a lease, even if the time is current.

tschottdorf commented:
Reviewed 1 of 1 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7139)*
<!-- Sent from Reviewable.io -->
